libtransmission/rpcimpl.cc

```diff
@@ -1696,6 +1706,115 @@ static char const* torrentAdd(tr_session* session, tr_variant* args_in, tr_varia
 ****
 ***/
 
+static char const* groupGet(tr_session* s, tr_variant* args_in, tr_variant* args_out, struct tr_rpc_idle_data* /*idle_data*/)
+{
+    tr_variant* names = nullptr;
+    std::string_view name;
+    int names_count = 0;
+
+    if (tr_variantDictFindStrView(args_in, TR_KEY_name, &name))
+    {
+        names_count = -1;
+    }
+    else if (tr_variantDictFindList(args_in, TR_KEY_name, &names))
+    {
+        names_count = tr_variantListSize(names);
+    }
+
+    tr_variant* list = tr_variantDictAddList(args_out, TR_KEY_group, 1);
+    tr_bandwidth_group* group = s->groups;
+    while (group != nullptr)
+    {
+        bool nameInList = false;
+        if (names_count > 0)
+        {
+            for (int i = 0; !nameInList && i < names_count; i++)
+            {
+                tr_variant* v = tr_variantListChild(names, i);
+                std::string_view l;
+                if (tr_variantIsString(v) && tr_variantGetStrView(v, &l))
+                {
+                    nameInList = group->name == l;
+                }
+            }
+        }
+
+        if (nameInList || (name == group->name) || (names_count == 0))
+        {
+            tr_variant* dict = nullptr;
+            bool u = false;
+            bool d = false;
+            uint32_t up = 0;
+            uint32_t down = 0;
+            dict = tr_variantListAddDict(list, 5);
+            tr_bandwidthGroupGetLimits(group, &u, &up, &d, &down);
+            tr_variantDictAddStrView(dict, TR_KEY_name, group->name);
+            tr_variantDictAddBool(dict, TR_KEY_uploadLimited, u);
+            tr_variantDictAddBool(dict, TR_KEY_downloadLimited, d);
+            tr_variantDictAddInt(dict, TR_KEY_downloadLimit, down);
+            tr_variantDictAddBool(dict, TR_KEY_honorsSessionLimits, group->bandwidth->areParentLimitsHonored(TR_UP));
+        }
+        group = group->next;
+    }
+
+    return nullptr;
+}
+
+static char const* groupSet(
+    tr_session* session,
+    tr_variant* args_in,
+    tr_variant* /*args_out*/,
+    struct tr_rpc_idle_data* /*idle_data*/)
+{
+    std::string_view name;
+    tr_bandwidth_group* group = nullptr;
+    bool u = false;
+    bool d = false;
+    bool honors = false;
+    uint32_t down = 0;
+    uint32_t up = 0;
+    int64_t intVal = 9;
```

Comment:
_(nit)_ move these declarations to where the variables are used, instead of frontloading everything to the beginning of the function